File: sections/71_discrete_convex_lagrangian/problems/b_exact_red_mst/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    it = iter(data)
    n, m, wanted = int(next(it)), int(next(it)), int(next(it))
    edges = []
    for _ in range(m):
        u, v, w, color = (
            int(next(it)) - 1,
            int(next(it)) - 1,
            int(next(it)),
            next(it),
        )
        edges.append((u, v, w, color == b"R"))

    def run(lam, prefer_red):
        order = sorted(
            edges,
            key=lambda e: (e[2] + lam * e[3], -e[3] if prefer_red else e[3]),
        )
        parent = list(range(n))
        size = [1] * n

        def find(x):
            while x != parent[x]:
                x = parent[x]
            return x

        used = red = value = 0
        for u, v, w, is_red in order:
            u, v = find(u), find(v)
            if u == v:
                continue
            if size[u] < size[v]:
                u, v = v, u
            parent[v] = u
            size[u] += size[v]
            used += 1
            red += is_red
            value += w + lam * is_red
        return used, red, value

    bound = 400_000_000_000_000
    used, minimum, _ = run(bound, False)
    _, maximum, _ = run(-bound, True)
    if used != n - 1 or not minimum <= wanted <= maximum:
        print("IMPOSSIBLE")
        return
    lo, hi = -bound, bound
    while lo < hi:
        mid = (lo + hi + 1) // 2
        if run(mid, True)[1] >= wanted:
            lo = mid
        else:
            hi = mid - 1
    _, _, modified = run(lo, True)
    print(modified - lo * wanted)


if __name__ == "__main__":
    main()

File: sections/71_discrete_convex_lagrangian/problems/b_exact_red_mst/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
struct DSU {
  vector<int> p, sz;
  DSU(int n) : p(n), sz(n, 1) { iota(p.begin(), p.end(), 0); }
  int find(int x) {
    while (x != p[x])
      x = p[x];
    return x;
  }
  bool unite(int a, int b) {
    a = find(a);
    b = find(b);
    if (a == b)
      return false;
    if (sz[a] < sz[b])
      swap(a, b);
    p[b] = a;
    sz[a] += sz[b];
    return true;
  }
};
struct Edge {
  int u, v, red;
  int64 w;
};
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, m, wanted;
  if (!(cin >> n >> m >> wanted))
    return 0;
  vector<Edge> edges(m);
  for (auto &e : edges) {
    char c;
    cin >> e.u >> e.v >> e.w >> c;
    --e.u;
    --e.v;
    e.red = c == 'R';
  }
  auto run = [&](int64 lambda, bool prefer_red) {
    vector<int> order(m);
    iota(order.begin(), order.end(), 0);
    sort(order.begin(), order.end(), [&](int i, int j) {
      int64 a = edges[i].w + lambda * edges[i].red;
      int64 b = edges[j].w + lambda * edges[j].red;
      if (a != b)
        return a < b;
      return prefer_red ? edges[i].red > edges[j].red
                        : edges[i].red < edges[j].red;
    });
    DSU dsu(n);
    int used = 0, red = 0;
    int64 value = 0;
    for (int id : order)
      if (dsu.unite(edges[id].u, edges[id].v)) {
        ++used;
        red += edges[id].red;
        value += edges[id].w + lambda * edges[id].red;
      }
    return tuple<int, int, int64>{used, red, value};
  };
  auto [u1, min_red, z1] = run(400000000000000LL, false);
  auto [u2, max_red, z2] = run(-400000000000000LL, true);
  if (u1 != n - 1 || wanted < min_red || wanted > max_red) {
    cout << "IMPOSSIBLE\n";
    return 0;
  }
  int64 lo = -400000000000000LL, hi = 400000000000000LL;
  while (lo < hi) {
    int64 mid = lo + (hi - lo + 1) / 2;
    auto [used, red, value] = run(mid, true);
    if (red >= wanted)
      lo = mid;
    else
      hi = mid - 1;
  }
  auto [used, red, modified] = run(lo, true);
  cout << modified - lo * wanted << '\n';
}
